### PR TITLE
database: Make wrapped version of timed_out_error a timed_out_error

### PR DESCRIPTION
**Pull request by elcallio**
Refs #9919

in a6202ae  throw_commitlog_add_error was added to ensure we had more
info on errors generated writing to commit log.

However, several call sites catch timed_out_error explicitly, not
checking for nested etc.
97bb1be and 868b572 tried to deal with it, by using check routines.
It turns out there are call sites left, and while these should be
changed, it is safer and quicker for now to just ensure that
iff we have a timed_out_error, we throw yet another timed_out_error.

### CODE REVIEW

File: replica/database.cc [hunk 1]
```diff
@@ -1723,10 +1723,31 @@ future<mutation> database::apply_counter_update(schema_ptr s, const frozen_mutat
...
+// catching timed_out_error (not checking is_timeout_exception).
+// Fixing the call sites is a good idea, but it is also hard
+// to verify. This workaround should ensure we take the
+// correct code paths in all cases, until we can clean things up
+// proper.
+class wrapped_timed_out_error : public timed_out_error {
+private:
+    sstring _msg;
+public:
+    wrapped_timed_out_error(sstring msg)
```
Comment (bhalevy) [inline replica/database.cc:1737]: noexcept

File: replica/database.cc [hunk 2]
```diff
@@ -1723,10 +1723,31 @@ future<mutation> database::apply_counter_update(schema_ptr s, const frozen_mutat
...
 }
 
+// #9919 etc. The initiative to wrap exceptions here
+// causes a bunch of problems with (implicit) call sites
+// catching timed_out_error (not checking is_timeout_exception).
+// Fixing the call sites is a good idea, but it is also hard
+// to verify. This workaround should ensure we take the
+// correct code paths in all cases, until we can clean things up
+// proper.
+class wrapped_timed_out_error : public timed_out_error {
```
Comment (bhalevy) [inline replica/database.cc:1733]: maybe derive from both timed_out_error and std::runtime_error?

File: replica/database.cc [hunk 3]
```diff
@@ -1723,10 +1723,31 @@ future<mutation> database::apply_counter_update(schema_ptr s, const frozen_mutat
...
+    }
+};
+
+// see above (#9919)
+template<typename T = std::runtime_error>
 static void throw_commitlog_add_error(schema_ptr s, const frozen_mutation& m) {
     // it is tempting to do a full pretty print here, but the mutation is likely
     // humungous if we got an error, so just tell us where and pk...
-    std::throw_with_nested(std::runtime_error(format("Could not write mutation {}:{} ({}) to commitlog"
+    std::throw_with_nested(T(format("Could not write mutation {}:{} ({}) to commitlog"
```
Comment (bhalevy) [inline replica/database.cc:1750]: What's the point of `throw_with_nested`?
Why is it needed?

File: replica/database.cc [hunk 4]
```diff
@@ -1739,8 +1760,11 @@ future<> database::apply_with_commitlog(column_family& cf, const mutation& m, db
         try {
             commitlog_entry_writer cew(m.schema(), fm, db::commitlog::force_sync::no);
             h = co_await cf.commitlog()->add_entry(m.schema()->id(), cew, timeout);
+        } catch (timed_out_error&) {
+            // see above (#9919)
+            throw_commitlog_add_error<wrapped_timed_out_error>(cf.schema(), fm);
         } catch (...) {
-            throw_commitlog_add_error(cf.schema(), fm);
+            throw_commitlog_add_error<>(cf.schema(), fm);
```
Comment (bhalevy) [inline replica/database.cc:1767]: I think it would be clearer to have no default for the template parameter
and pass std::runtime_error here explicitly.

**Review by bhalevy — CHANGES_REQUESTED** [review]

File: replica/database.cc (hunk 2 above)
Comment (elcallio) [reply replica/database.cc:1733]: Both derive (non-virtual) from std::exception -> not a good idea.

File: replica/database.cc (hunk 3 above)
Comment (elcallio) [reply replica/database.cc:1750]: The point is we want to know the table and pk, tracing back to older issues with errors writing but no log info on who/what wrote. We could log here, but it is cleaner to allow caller to either log or handle things. Unfortunately we broke some catch points.

File: replica/database.cc (hunk 4 above)
Comment (elcallio) [reply replica/database.cc:1767]: Sure.

**Comment by nyh** [comment]
What does "Refs #9919" mean? If it fixes the issue, please use the word "fixes". If it doesn't, can you say a few words on what problem remains? Or that it maybe fixes 9919 but you're just not sure?

File: replica/database.cc (hunk 2 above)
Comment (bhalevy) [reply replica/database.cc:1733]: okay

File: replica/database.cc (hunk 3 above)
Comment (bhalevy) [reply replica/database.cc:1750]: But the table and pk are printed into the exception message, why do need to throw it as a nested exception?

**Comment by nyh** [comment]
> What does "Refs #9919" mean? If it fixes the issue, please use the word "fixes". If it doesn't, can you say a few words on what problem remains? Or that it maybe fixes 9919 but you're just not sure?

I think I can answer myself - IIUC, this patch should make it easier to debug #9919, but not fix anything in the underlying bug. Right?

**Comment by elcallio** [comment]
To be honest, right now we don't know if this fixes or not. Because it could be that the timeouts in #9919 (et al) are "expected" (they normally happen), but don't show because as long as RPC calls send the right error, things back off higher up in the app, and all is "well". It could also be that the timeouts are _not_ expected, and we somehow timeout _more_ than before, in which case this at least should move us closer to knowing so...

**Pull request closed without merging**